libraries/tools/kotlin-maven-plugin/src/main/java/org/jetbrains/kotlin/maven/KotlinCompileMojoBase.java

```diff
@@ -133,15 +138,32 @@ public void execute() throws MojoExecutionException, MojoFailureException {
         configureCompilerArguments(arguments, compiler);
         printCompilerArgumentsIfDebugEnabled(arguments, compiler);
 
-        MessageCollector messageCollector = new MavenPluginLogMessageCollector(getLog());
+        MavenPluginLogMessageCollector messageCollector = new MavenPluginLogMessageCollector(getLog(), true);
 
         ExitCode exitCode = compiler.exec(messageCollector, Services.EMPTY, arguments);
 
         switch (exitCode) {
             case COMPILATION_ERROR:
-                throw new MojoExecutionException("Compilation error. See log for more details");
             case INTERNAL_ERROR:
-                throw new MojoExecutionException("Internal compiler error. See log for more details");
+                throw new KotlinCompilationFailureException(
+                        CollectionsKt.map(messageCollector.getCollectedErrors(), new Function1<Pair<CompilerMessageLocation, String>, CompilerMessage>() {
+                            @Override
+                            public CompilerMessage invoke(Pair<CompilerMessageLocation, String> pair) {
+                                String lineContent = pair.getFirst().getLineContent();
+                                int lineContentLength = lineContent == null ? 0 : lineContent.length();
+
+                                return new CompilerMessage(
+                                        pair.getFirst().getPath(),
+                                        CompilerMessage.Kind.ERROR,
+                                        pair.getFirst().getLine(),
```

Comment:
-1 should be 0 so CompileMessage will handle it properly. I've extracted a function `fixLocation` to handle it.